### PR TITLE
ARROW-12953: [C++][Compute] Refactor CheckScalar* to take Datum arguments

### PR DESCRIPTION
**Pull request by diegodfrf**
Update shared_ptr<Scalar> and shared_ptr<Arrow> to Datum in CheckScalar* functions

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/scalar_string_test.cc [hunk 1]
```diff
@@ -1186,9 +1186,9 @@ TYPED_TEST(TestStringKernels, BinaryJoin) {
   auto expected =
       ArrayFromJSON(this->type(), R"(["a--bb--ccc", "", null, "dd", null, "ff--"])");
   CheckScalarBinary("binary_join", ArrayFromJSON(list(this->type()), list_json),
-                    separator, expected);
+                    static_cast<Datum>(separator), expected);
   CheckScalarBinary("binary_join", ArrayFromJSON(large_list(this->type()), list_json),
-                    separator, expected);
+                    static_cast<Datum>(separator), expected);
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_string_test.cc:1191]: Ah, this isn't a cast, but is instead invoking the implicit conversion ([case (1) here](https://en.cppreference.com/w/cpp/language/static_cast)). Can we write it as just `Datum(separator)` to make that clear?

Comment (diegodfrf) [reply cpp/src/arrow/compute/kernels/scalar_string_test.cc]: Done

**Comment by edponce** [comment]
LGTM. Thanks for working on this!

**Review by edponce — APPROVED** [review]

**Comment by lidavidm** [comment]
R failure is unrelated, merging. Thanks @diegodfrf!

**Pull request closed without merging**